In [251]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as mpt
import seaborn as sns

## 1. Chargement et structure

**Question** : Quelle est la structure du dataset ?

In [3]:
df = pd.read_csv('../data/raw/Sample - Superstore.csv')
df.shape

(10064, 21)

In [4]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

**Observation** : 10 064 lignes, 21 colonnes. Types : 1 int (Row ID),
4 float (Sales, Quantity, Discount, Profit), 16 object (texte/dates en string).

**Décision** : les colonnes de dates (`Order Date`, `Ship Date`) et numériques
en string (`Postal Code`) devront être reconverties dans la phase de nettoyage.

In [256]:
df.dtypes

Row ID             int64
Order ID          object
Order Date        object
Ship Date         object
Ship Mode         object
Customer ID       object
Customer Name     object
Segment           object
Country           object
City              object
State             object
Postal Code       object
Region            object
Product ID        object
Category          object
Sub-Category      object
Product Name      object
Sales            float64
Quantity         float64
Discount         float64
Profit           float64
dtype: object

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10064 entries, 0 to 10063
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         10064 non-null  int64  
 1   Order ID       10064 non-null  object 
 2   Order Date     10064 non-null  object 
 3   Ship Date      9963 non-null   object 
 4   Ship Mode      9763 non-null   object 
 5   Customer ID    10064 non-null  object 
 6   Customer Name  9713 non-null   object 
 7   Segment        10064 non-null  object 
 8   Country        10064 non-null  object 
 9   City           10064 non-null  object 
 10  State          10064 non-null  object 
 11  Postal Code    9864 non-null   object 
 12  Region         10064 non-null  object 
 13  Product ID     10064 non-null  object 
 14  Category       10064 non-null  object 
 15  Sub-Category   10064 non-null  object 
 16  Product Name   10064 non-null  object 
 17  Sales          9863 non-null   float64
 18  Quanti

**Observation** : `Order Date` et `Ship Date` sont des `object` alors qu'elles
représentent des dates. `Postal Code` est un `object` alors qu'il devrait être numérique.

**Décision** : conversion en `datetime` et `int` prévue dans le nettoyage.

df.head()

In [8]:
df.tail()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
10059,1281,CA-2016-160815,9/5/2016,9/6/2016,First Class,TR-21325,TOBY RITTER,Consumer,United States,Cedar Rapids,...,52402.0,Central,TEC-PH-10003505,Technology,Phones,Geemarc AmpliPOWER60,278.400,3.0,0.0,80.7360
10060,7699,CA-2017-151799,12/14/2017,12/18/2017,Standard Class,BF-11170,BEN FERRER,Home Office,United States,Lawrence,...,1841.0,East,OFF-ST-10002790,Office Supplies,Storage,Safco Industrial Shelving,73.850,1.0,0.0,2.2155
10061,8597,CA-2014-111934,5/5/2014,5/7/2014,First Class,GD-14590,GIULIETTA DORTCH,Corporate,United States,Arlington,...,22204.0,South,OFF-PA-10000474,Office Supplies,Paper,Easy-staple paper,35.440,1.0,0.0,16.6568
10062,4962,CA-2014-156587,3/7/2014,3/8/2014,First Class,AB-10015,NaN,Consumer,United States,Seattle,...,98103.0,West,FUR-CH-10004477,Furniture,Chairs,"Global Push Button Manager's Chair, Indigo",48.712,1.0,0.2,5.4801
10063,3998,CA-2015-105627,3/8/2015,3/12/2015,Standard Class,DK-12895,DANA KAYDOS,Consumer,United States,Kenosha,...,53142.0,Central,OFF-AR-10002704,Office Supplies,Art,Boston 1900 Electric Pencil Sharpener,14.980,1.0,0.0,4.4940


In [258]:
df.describe(include='all')
#il y a du manque de valeurs pour sales et quatites

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
count,10030.000000,10030,10030,9929,9729,10030,9713,10030,10030,10030,...,9831,10030,10030,10030,10030,10030,9.830000e+03,9831.000000,10030.000000,10030.000000
unique,NaN,5009,1495,1334,4,793,810,6,1,633,...,705,4,1862,6,17,1850,NaN,NaN,NaN,NaN
top,NaN,CA-2017-100111,9/5/2016,12/16/2015,Standard Class,WB-21850,William Brown,Consumer,United States,New York City,...,10035.0,West,OFF-PA-10001970,Office Supplies,Binders,Easy-staple paper,NaN,NaN,NaN,NaN
freq,NaN,14,39,35,5800,38,36,5148,10030,896,...,259,3215,19,5927,1529,48,NaN,NaN,NaN,NaN
mean,4996.756231,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,1.956111e+03,3.780389,0.157536,28.609792
std,2884.870357,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,4.418059e+04,2.240859,0.210657,233.859007
min,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,4.440000e-01,-5.000000,0.000000,-6599.978000
25%,2499.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,1.728000e+01,2.000000,0.000000,1.731375
50%,4995.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,5.449000e+01,3.000000,0.200000,8.652800
75%,7495.750000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,2.106550e+02,5.000000,0.200000,29.341200


In [282]:
df.isnull().sum()

Row ID             0
Order ID           0
Order Date         0
Ship Date        101
Ship Mode        301
Customer ID        0
Customer Name    317
Segment            0
Country            0
City               0
State              0
Postal Code      199
Region             0
Product ID         0
Category           0
Sub-Category       0
Product Name       0
Sales            200
Quantity         199
Discount           0
Profit             0
Order Date dt    301
Ship Date dt     101
dtype: int64

In [11]:
df.duplicated().value_counts()
#34 duplicated row

False    10030
True        34
Name: count, dtype: int64

**Observation** : 34 lignes dupliquées (0,3%).

**Décision** : supprimer les doublons via `drop_duplicates()` au nettoyage.

In [13]:
pattern = r'^\d{2}/\d{2}/\d{4}$'
r = df['Ship Date'].str.match(pattern, na=False).value_counts()
r

Ship Date
False    7501
True     2563
Name: count, dtype: int64

In [14]:
df.nunique()
#le nombre d'elements unique pour choisir le bonne type d'necodage (categorielle)

Row ID           9994
Order ID         5009
Order Date       1495
Ship Date        1334
Ship Mode           4
Customer ID       793
Customer Name     810
Segment             6
Country             1
City              633
State              86
Postal Code       705
Region              4
Product ID       1862
Category            6
Sub-Category       17
Product Name     1850
Sales            5738
Quantity           15
Discount           13
Profit           7287
dtype: int64

**Observations** :

| Colonne | Uniques | Interprétation |
|---|---|---|
| Row ID | 9994 | Identifiant quasi unique |
| Order ID | 5009 | Clé de regroupement |
| Customer ID | 793 | Clé client |
| Product ID | 1862 | Clé produit |
| Country | 1 | **Colonne constante** → à supprimer |
| Category | 6 | Faible cardinalité → encodage one-hot |
| Sub-Category | 17 | Faible cardinalité |
| State | 86 | Cardinalité moyenne |

**Décision** :
- `Country` : à supprimer (valeur constante)
- `Category`, `Segment`, `Region`, `Ship Mode` : one-hot encoding
- `Customer ID`, `Product ID`, `Order ID` : identifiants à garder tels quels

In [108]:
#Missiong value :
df.isna().sum().sort_values(ascending=False)
missing = pd.DataFrame({
    "missing_count" : df.isna().sum(),
    "missing_pencent" : df.isna().mean() *100
})
missing = missing[missing['missing_count'] > 0]
missing.sort_values('missing_count', ascending=False)

,missing_count,missing_pencent
Customer Name,317,3.160518
Ship Mode,301,3.000997
Sales,200,1.994018
Postal Code,199,1.984048
Quantity,199,1.984048
Ship Date,101,1.006979


In [160]:

# Customer Name
df.loc[df["Customer Name"].isna(), 'Customer ID'] #246
customer_check = (
    df.groupby('Customer ID')['Customer Name'].agg(['count', 'first'])
)
customer_check.head(10)
customer_check["first"].isna().sum()
#j'ai regrouper les lignes par Customer Id , puis j'ai pris le premier Customer Name trouve pour chaque Client, et compter combien de Customer ID differnts ont un Customer Name manquat j'ai trouver 1
customer_check[customer_check['first'].isna()] # LD-16855	0	None
#Il ua un client Customer Id pour lequel aucun non n'est disponible dans les donnees
customer_idmissing = customer_check[customer_check['first'].isna()].index
df[df["Customer ID"].isin(customer_idmissing)][
    ["Order ID", "Customer ID", "Customer Name"]
]
#df.loc[df['Customer ID'].isin(customer_idmissing)] # CA-2016-152331	
#df.loc[df['Customer ID'] =='LD-16855', ['Order ID', 'Customer Name']]
#r = df.groupby('Order ID')[['Customer Name', 'Customer ID']].agg(['count', 'first'])
#r.loc[r['Order ID'] == "CA-2016-152331"]
# => Aucun CUstomer Name disponible et ce Customer ID a un

,Order ID,Customer ID,Customer Name
6505,CA-2016-152331,LD-16855,NaN


In [167]:
df[df["Customer ID"] == 'LD-16855'].shape

(1, 21)

In [168]:
df[df["Customer ID"] == 'LD-16855'][
    ["Order ID", "Customer ID", "Customer Name"]
]
# Comme decision on les supprimer car il est MCAR ET SONT POURCENTAGE < 5%
# Customer Name
# Les valeurs manquantes peuvent être reconstruites grâce au Customer ID.
# Un même Customer ID correspond à un même client.
# Nous utilisons donc les autres lignes du même client pour récupérer son nom.

,Order ID,Customer ID,Customer Name
6505,CA-2016-152331,LD-16855,NaN


In [166]:
#df_clean = df.copy()
#
#df_clean["Customer Name"] = (
#    df_clean.groupby("Customer ID")["Customer Name"]
#    .transform("first")
#)
#df_clean[df_clean["Customer Name"].isna()]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
6505,6506,CA-2016-152331,6/26/2016,6/30/2016,Standard Class,LD-16855,None,Corporate,United States,Chicago,...,60653.0,Central,OFF-AR-10001547,Office Supplies,Art,Newell 311,5.304,3.0,0.2,0.4641


In [77]:
df['Segment'].value_counts()
#tester les fautes d'orthographes



Postal Code
10035.0      259
10009.0      227
10024.0      223
94122.0      194
10011.0      188
            ... 
34741.0        1
94109.0.0      1
30080.0.0      1
6810.0         1
72762.0        1
Name: count, Length: 705, dtype: int64

In [88]:
OrderDate = pd.to_datetime(df['Order Date'], errors='coerce')
ShapeDate = pd.to_datetime(df['Ship Date'], errors='coerce')
OrderDate.isna().value_counts()
# lors du casting de type de Order Date il aura des dates invalides(envirant de 301) apres le casting
#Pour le Ship Date il ya 101 nan avant et apres le casting

Order Date
False    9729
True      301
Name: count, dtype: int64

In [101]:
r = df['Customer Name'].str.replace(r'\s+', '', regex=True).str.lower()
r.isna().value_counts()

Customer Name
False    9713
True      317
Name: count, dtype: int64

In [169]:
# Ship Mode
df['Ship Mode'].isna().value_counts()
# 301 valeurs manquantes

Ship Mode
False    9729
True      301
Name: count, dtype: int64

In [176]:
df[df['Ship Mode'].isna()].head(20)
#df[df['Ship Mode'].notna()].head(20)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
28,29,US-2015-150630,9/17/2015,9/21/2015,NaN,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,19140.0,East,OFF-BI-10000474,Office Supplies,Binders,Avery Recycled Flexi-View Covers for Binding S...,9.618,2.0,0.7,-7.0532
94,95,CA-2015-149587,1/31/2015,2/5/2015,NaN,KB-16315,Karl Braun,Consumer,United States,Minneapolis,...,55407.0,Central,OFF-BI-10002852,Office Supplies,Binders,Ibico Standard Transparent Covers,32.960,2.0,0.0,16.1504
191,192,CA-2015-102281,10/12/2015,10/14/2015,NaN,MP-17470,Mark Packer,Home Office,United States,New York City,...,10035.0,East,OFF-PA-10000061,Office Supplies,Paper,Xerox 205,51.840,8.0,0.0,24.8832
252,253,CA-2016-146941,12/10/2016,12/13/2016,NaN,DL-13315,Delfina Latchford,Consumer,United States,New York City,...,10024.0,East,OFF-ST-10001228,Office Supplies,Storage,"Fellowes Personal Hanging Folder Files, Navy",80.580,6.0,0.0,22.5624
314,315,CA-2014-167850,8/9/2014,8/16/2014,NaN,AG-10525,Andy Gerbode,Corporate,United States,Saint Petersburg,...,33710.0,South,TEC-PH-10002398,Technology,Phones,AT&T 1070 Corded Phone,178.384,2.0,0.2,22.2980
364,365,CA-2017-144904,9/25/2017,10/1/2017,NaN,KW-16435,Katrina Willman,Consumer,United States,New York City,...,10009.0,East,FUR-FU-10000023,Furniture,Furnishings,Eldon Wave Desk Accessories,47.120,8.0,0.0,20.7328
385,386,US-2015-168935,11/27/2015,12/2/2015,NaN,DO-13435,Denny Ordway,Consumer,United States,Pembroke Pines,...,NaN,South,TEC-AC-10002335,Technology,Accessories,Logitech Media Keyboard K200,83.976,3.0,0.2,-1.0497
482,483,CA-2014-127691,7/22/2014,7/27/2014,NaN,EM-14065,Erin Mull,Consumer,United States,New York City,...,10024.0,East,OFF-AR-10002053,Office Supplies,Art,"Premium Writing Pencils, Soft, #2 by Central A...",5.960,2.0,0.0,1.6688
483,484,CA-2014-127691,7/22/2014,7/27/2014,NaN,EM-14065,NaN,Consumer,United States,New York City,...,10024.0,East,TEC-AC-10002567,Technology,Accessories,Logitech G602 Wireless Gaming Mouse,159.980,2.0,0.0,57.5928
502,503,US-2016-105578,5/30/2016,6/4/2016,NaN,MY-17380,Maribeth Yedwab,Corporate,United States,Parker,...,80134.0,West,FUR-CH-10001215,Furniture,Chairs,Global Troy Executive Leather Low-Back Tilter,801.568,2.0,0.2,50.0980


In [177]:
df[df["Ship Mode"].isna()][
    ["Order ID", "Order Date", "Ship Date", "Customer ID"]
].head(20)
# il ya des dates invalides dasn order date et ship date aussi

,Order ID,Order Date,Ship Date,Customer ID
28,US-2015-150630,9/17/2015,9/21/2015,TB-21520
94,CA-2015-149587,1/31/2015,2/5/2015,KB-16315
191,CA-2015-102281,10/12/2015,10/14/2015,MP-17470
252,CA-2016-146941,12/10/2016,12/13/2016,DL-13315
314,CA-2014-167850,8/9/2014,8/16/2014,AG-10525
364,CA-2017-144904,9/25/2017,10/1/2017,KW-16435
385,US-2015-168935,11/27/2015,12/2/2015,DO-13435
482,CA-2014-127691,7/22/2014,7/27/2014,EM-14065
483,CA-2014-127691,7/22/2014,7/27/2014,EM-14065
502,US-2016-105578,5/30/2016,6/4/2016,MY-17380


In [183]:
df[df['Ship Mode'].isna()]['Order ID'].map(
    df.groupby('Order ID')['Ship Mode'].first()
).notna().sum()
#j'observe que pour les valeurs manquantes leur Order Id se repete donc nous pouvant le reconstruire d'apres cela 

np.int64(231)

In [205]:
#Ship Date
df[df["Ship Date"].isna()][
    ["Order ID", "Order Date", "Ship Date", "Ship Mode"]
].groupby('Order ID')['Ship Date'].count().sort_values( ascending=False)

Order ID
CA-2014-104269    0
CA-2014-107139    0
CA-2014-107524    0
CA-2014-113383    0
CA-2014-115084    0
                 ..
US-2017-120418    0
US-2017-147221    0
US-2017-153948    0
US-2017-162558    0
US-2017-162670    0
Name: Ship Date, Length: 101, dtype: int64

In [200]:
df.sort_values('Order ID').head(30)
df.groupby('Order ID')['Ship Date'].count().head(20)
# j'observe que pour les valeurs manquantes leur Order Id se repete donc nous pouvant le reconstruire d'apres cela 

Order ID
CA-2014-100006    1
CA-2014-100090    2
CA-2014-100293    1
CA-2014-100328    1
CA-2014-100363    2
CA-2014-100391    1
CA-2014-100678    4
CA-2014-100706    2
CA-2014-100762    4
CA-2014-100860    1
CA-2014-100867    1
CA-2014-100881    1
CA-2014-100895    3
CA-2014-100916    3
CA-2014-100972    1
CA-2014-101147    1
CA-2014-101175    1
CA-2014-101266    1
CA-2014-101364    1
CA-2014-101392    1
Name: Ship Date, dtype: int64

In [246]:
shipmode_check = (
    df.groupby("Order ID")[["Ship Mode", 'Ship Date', 'Order Date']]
      .agg(["count", "nunique", "first"])
)

shipmode_check.head(20)
# voir les nombres de valeurs non manquates avec la premier val trouve

Ship Mode                         Ship Date          \
                   count nunique           first     count nunique   
Order ID                                                             
CA-2014-100006         1       1  Standard Class         1       1   
CA-2014-100090         2       1  Standard Class         2       1   
CA-2014-100293         1       1  Standard Class         1       1   
CA-2014-100328         1       1  Standard Class         1       1   
CA-2014-100363         1       1  Standard Class         2       1   
CA-2014-100391         1       1  Standard Class         1       1   
CA-2014-100678         4       1  Standard Class         4       1   
CA-2014-100706         2       1    Second Class         2       1   
CA-2014-100762         3       1  Standard Class         4       1   
CA-2014-100860         1       1    Second Class         1       1   
CA-2014-100867         1       1  Standard Class         1       1   
CA-2014-100881         1       1  Standard Class         1       1   
CA-2014-100895         3       1  Standard Class         3       1   
CA-2014-100916         3       1  Standard Class         3       1   
CA-2014-100972         1       1    Second Class         1       1   
CA-2014-101147         1       1     First Class         1       1   
CA-2014-101175         1       1  Standard Class         1       1   
CA-2014-101266         1       1    Second Class         1       1   
CA-2014-101364         1       1  Standard Class         1       1   
CA-2014-101392         1       1  Standard Class         1       1   

                           Order Date                      
                     first      count nunique       first  
Order ID                                                   
CA-2014-100006   9/13/2014          1       1    9/7/2014  
CA-2014-100090   7/12/2014          2       1    7/8/2014  
CA-2014-100293   3/18/2014          1       1   3/14/2014  
CA-2014-100328    2/3/2014          1       1   1/28/2014  
CA-2014-100363   4/15/2014          2       1    4/8/2014  
CA-2014-100391   5/29/2014          1       1   5/25/2014  
CA-2014-100678   4/22/2014          4       1   4/18/2014  
CA-2014-100706  12/18/2014          2       1  12/16/2014  
CA-2014-100762  11/29/2014          4       1  11/24/2014  
CA-2014-100860   3/30/2014          1       1   3/26/2014  
CA-2014-100867  10/24/2014          1       1  10/19/2014  
CA-2014-100881    4/1/2014          1       1   3/28/2014  
CA-2014-100895    6/6/2014          3       1    6/2/2014  
CA-2014-100916  10/26/2014          3       1  10/21/2014  
CA-2014-100972  11/24/2014          1       1  11/19/2014  
CA-2014-101147   12/4/2014          1       1   12/2/2014  
CA-2014-101175  12/14/2014          1       1   12/9/2014  
CA-2014-101266   8/30/2014          1       1   8/27/2014  
CA-2014-101364  12/26/2014          1       1  12/22/2014  
CA-2014-101392  12/13/2014          1       1   12/7/2014

In [244]:
shipmode_missing_orders = shipmode_check[
    shipmode_check["count"] < df.groupby("Order ID").size()
]

shipmode_missing_orders

,count,nunique,first
Order ID,,,
CA-2014-100363,1,1,Standard Class
CA-2014-100762,3,1,Standard Class
CA-2014-103331,2,1,Standard Class
CA-2014-103373,0,0,None
CA-2014-107181,1,1,Standard Class
...,...,...,...
US-2017-139465,1,1,First Class
US-2017-147886,6,1,First Class
US-2017-154872,2,1,Standard Class


In [209]:
df[df["Ship Mode"].isna()][
    ["Order ID", "Ship Mode"]
].sort_values("Order ID")

,Order ID,Ship Mode
3836,CA-2014-100363,NaN
6317,CA-2014-100762,NaN
5333,CA-2014-103331,NaN
1829,CA-2014-103373,NaN
1544,CA-2014-107181,NaN
...,...,...
1797,US-2017-139465,NaN
5743,US-2017-147886,NaN
6544,US-2017-154872,NaN
5013,US-2017-160465,NaN


In [243]:
df_copy = df.copy()
order_shipmode = (
    df_copy.groupby("Order ID")["Ship Mode"]
      .agg(["count", "nunique", "first"])
)

r = df_copy[
    df_copy["Ship Mode"].isna()
].merge(
    order_shipmode,
    left_on="Order ID",
    right_index=True
)[
    ["Order ID", "Ship Mode", "count", "nunique", "first"]
]
#df_copy['Ship Mode'] = r[df_copy['Order ID'] == r['Order ID']] 
#r.loc[r['Order ID'] == 'CA-2014-100363']
#df.loc[df['Order ID'] == 'CA-2014-100363', 'Ship Mode']
df_copy['Ship Mode'] = df.groupby('Order ID')['Ship Mode'].transform('first')
df_copy[df_copy['Ship Mode'].isna()].groupby('Order ID')['Ship Mode'].agg(["count", "nunique", "first"])['count'].sort_values(ascending=False)
df_copy["Ship Mode"].isna().sum()
#j'ai deja les transformer maintent me reste 70 qui n'ont pas un order id qui se repete, donc ici nous avons voir le order Date et le ship Date
#Ship Mode : 301 valeurs manquantes ont été détectées. Une analyse par Order ID a permis de récupérer 231 valeurs lorsque le même Order ID possédait une autre ligne avec un Ship Mode connu. Les 70 valeurs restantes ne disposaient pas d'une information suffisante dans le dataset pour être reconstruites de manière fiable.
#ont été imputés par le mode global "Standard Class" (57,8% du dataset), avec la limite assumée que cette imputation introduit un biais.

np.int64(70)

In [241]:
df_copy[df_copy["Ship Mode"].isna()][
    ["Order ID", "Ship Mode", 'Product ID','Order Date','Ship Date',]
].sort_values("Order ID").head()

,Order ID,Ship Mode,Product ID,Order Date,Ship Date
1829,CA-2014-103373,None,TEC-PH-10002885,5/18/2014,5/24/2014
482,CA-2014-127691,None,OFF-AR-10002053,7/22/2014,7/27/2014
483,CA-2014-127691,None,TEC-AC-10002567,7/22/2014,7/27/2014
5220,CA-2014-140487,None,FUR-BO-10000711,6/14/2014,6/20/2014
7397,CA-2014-141649,None,OFF-AP-10003217,9/30/2014,9/30/2014


In [219]:
df.groupby("Order ID")["Order Date"].nunique().value_counts()

Order Date
1    4782
2     227
Name: count, dtype: int64

In [220]:
df.groupby("Order ID")["Ship Date"].nunique().value_counts()

Ship Date
1    4981
0      23
2       5
Name: count, dtype: int64

In [252]:
sns.barplot(data=df, x="Ship Mode", y="Ship Date")

<Axes: title={'center': 'Ship Mode = Same Day'}, xlabel='Ship Mode', ylabel='Ship Date'>

In [255]:
sns.scatterplot(df, x='Order ID', y='Category')

<Axes: title={'center': 'Ship Mode = Same Day'}, xlabel='Ship Mode', ylabel='Ship Date'>

Les colonnes Order Date et Ship Date sont stockées en object (string) et présentent deux formats : M/D/YYYY (majoritaire) et YYYY-MM-DD (rare). Une conversion pd.to_datetime avec errors='coerce' a été appliquée. 301 dates non-parsables ont été détectées et traitées comme des valeurs manquantes.

Sales = Quantity × Prix_unitaire × (1 - Discount)
supprimer les 200 lignes avec Sales manquant. 200 lignes sur 10 064 = 2% seulement, Ces lignes n'ont ni Sales ni Quantity ni souvent Ship Date → elles sont probablement incomplètes globalement

Une analyse sans elles reste valide
ou bien  imputer Sales par la médiane des Sales du même produit.

In [285]:

df.drop_duplicates()
df['Category'].value_counts()

Category
Office Supplies    5927
Furniture          2085
Technology         1817
office supplies     123
furniture            42
technology           36
Name: count, dtype: int64

In [261]:
df['Segment'].value_counts()


Segment
Consumer       5148
Corporate      3003
Home Office    1778
Consumerr        58
Corporrate       26
Home Ofice       17
Name: count, dtype: int64

In [262]:
df['Region'].value_counts()


Region
West       3215
East       2855
Central    2334
South      1626
Name: count, dtype: int64

In [263]:
df['Ship Mode'].value_counts()

Ship Mode
Standard Class    5800
Second Class      1891
First Class       1508
Same Day           530
Name: count, dtype: int64

Les colonnes catégorielles (Category, Segment, Region, Ship Mode) ont été inspectées via value_counts(). Aucune faute d'orthographe détectée dans Region, Ship Mode ms pour Segment et Category il faut les coriger en transformant tout en lower case et les regroupes. Une normalisation (trim des espaces, harmonisation de la casse) sera appliquée par précaution.

Postal Code présente 200 valeurs manquantes. Une imputation via City a été envisagée : analyse de la cardinalité City → Postal Code montre que X% des villes ont un code postal unique. Décision : imputer les codes.

In [264]:
#Cardinalité City → Postal Code
df.groupby('City')['Postal Code'].nunique().value_counts()

Postal Code
1    516
2     70
3     24
5      7
0      5
4      4
6      4
8      2
7      1
Name: count, dtype: int64

L'imputation de Postal Code via City a été testée : 516 villes sur 633 (81%) ont un code postal unique, permettant une imputation fiable. Pour les 117 villes à codes multiples ou sans code, la valeur a été marquée "UNKNOWN" pour éviter l'introduction de bruit. X valeurs ont été imputées, Y restent "UNKNOWN".

In [265]:
#Lien Sales manquant ↔ Quantity manquant
df[df['Sales'].isna()]['Quantity'].isna().value_counts()
# => supprimer les lignes

Quantity
False    196
True       4
Name: count, dtype: int64

Une hypothèse de reconstruction de Sales via Quantity et Product Name a été testée. Résultat : 196 des 200 lignes avec Sales manquant ont également Quantity manquant (98%). Ces lignes sont globalement incomplètes et le prix unitaire n'est pas constant dans le dataset. Décision : suppression des 200 lignes.

In [266]:
df['Category'].value_counts()


Category
Office Supplies    5927
Furniture          2085
Technology         1817
office supplies     123
furniture            42
technology           36
Name: count, dtype: int64

La colonne Category présente 6 valeurs distinctes au lieu de 3 attendues. Analyse : 201 lignes contiennent les variantes minuscules "office supplies", "furniture", "technology". Décision : normalisation de la casse via .str.title() pour harmoniser à 3 modalités.

In [267]:
df['Segment'].value_counts()


Segment
Consumer       5148
Corporate      3003
Home Office    1778
Consumerr        58
Corporrate       26
Home Ofice       17
Name: count, dtype: int64

La colonne Segment présente 6 valeurs distinctes au lieu de 3 attendues. Analyse : 101 lignes contiennent des typos ("Consumerr", "Corporrate", "Home Ofice"). Décision : correction via une table de correspondance explicite (mapping manuel) car les typos ne peuvent pas être corrigées par une simple normalisation de casse.

In [275]:
df['Region'].value_counts()

Region
West       3215
East       2855
Central    2334
South      1626
Name: count, dtype: int64

Les colonnes Region et Ship Mode présentent respectivement 4 modalités propres, sans variante de casse ni typo. Aucun nettoyage nécessaire.

In [276]:
df['Ship Mode'].value_counts()

Ship Mode
Standard Class    5800
Second Class      1891
First Class       1508
Same Day           530
Name: count, dtype: int64

Détection des valeurs aberrantes

In [278]:
#Quantité négative
neg_qty = df[df['Quantity'] < 0]
print(f"Quantité < 0 : {len(neg_qty)} lignes")
neg_qty[['Order ID', 'Quantity', 'Sales', 'Discount']].head(10)

Quantité < 0 : 10 lignes


,Order ID,Quantity,Sales,Discount
439,CA-2017-157252,-5.0,207.846,0.1
1481,US-2016-108455,-5.0,33.870,0.0
2302,CA-2016-106558,-5.0,316.000,0.0
6511,CA-2017-167640,-5.0,23.880,0.0
7419,US-2016-114013,-5.0,552.560,0.2
7894,CA-2017-124744,-5.0,82.400,0.0
8730,CA-2015-103870,-5.0,72.744,0.2
9560,CA-2016-129280,-5.0,51.968,0.2
9662,CA-2016-160717,-5.0,477.600,0.2
9936,CA-2017-132955,-5.0,102.930,0.0


10 lignes détectées (incohérent physiquement)

In [279]:
# Remise supérieure à 100%
over_discount = df[df['Discount'] > 1.0]
print(f"Discount > 1 : {len(over_discount)} lignes")
over_discount[['Order ID', 'Discount', 'Sales', 'Profit']].head(10)

Discount > 1 : 10 lignes


,Order ID,Discount,Sales,Profit
3096,CA-2015-114468,1.5,5.768,-13.5548
3501,CA-2016-108616,1.5,25.920,7.7760
4551,US-2014-165862,1.5,351.216,4.3902
4710,CA-2017-138149,1.5,62.310,22.4316
6405,CA-2017-154074,1.5,569.568,7.1196
6457,CA-2014-110065,1.5,15.960,7.0224
6683,CA-2014-109134,1.5,20.040,8.8176
8814,CA-2015-144274,1.5,307.980,89.3142
9291,CA-2014-163650,1.5,34.500,15.5250
9352,CA-2017-148411,1.5,520.464,-14.8704


10 lignes détectées (impossible commercialement)

In [281]:
#Ship Date < Order Date
df['Order Date dt'] = pd.to_datetime(df['Order Date'], errors='coerce')
df['Ship Date dt'] = pd.to_datetime(df['Ship Date'], errors='coerce')

invalid_dates = df[df['Ship Date dt'] < df['Order Date dt']]
print(f"Ship Date < Order Date : {len(invalid_dates)} lignes")
invalid_dates

Ship Date < Order Date : 8 lignes


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Order Date dt,Ship Date dt
1562,1563,US-2017-102890,6/30/2017,1/1/2010,Same Day,SG-20470,Sheri Gordon,Consumer,United States,New York City,...,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.630,5.0,0.4,-295.9785,2017-06-30,2010-01-01
2517,2518,CA-2017-123134,5/2/2017,1/1/2010,Standard Class,DW-13585,Dorothy Wardle,Corporate,United States,Westfield,...,FUR-FU-10003975,Furniture,Furnishings,Eldon Advantage Chair Mats for Low to Medium P...,129.930,3.0,0.0,12.9930,2017-05-02,2010-01-01
3279,3280,CA-2014-102988,4/5/2014,1/1/2010,Second Class,GM-14695,NaN,Corporate,United States,Alexandria,...,OFF-AR-10000127,Office Supplies,Art,Newell 321,22.960,7.0,0.0,6.6584,2014-04-05,2010-01-01
4939,4940,CA-2014-157147,1/13/2014,1/1/2010,Standard Class,BD-11605,Brian Dahlen,Consumer,United States,San Francisco,...,OFF-AR-10003514,Office Supplies,Art,4009 Highlighters by Sanford,19.900,5.0,0.0,6.5670,2014-01-13,2010-01-01
7099,7100,CA-2015-124933,12/26/2015,1/1/2010,Second Class,DF-13135,David Flashing,Consumer,United States,New York City,...,OFF-PA-10003302,Office Supplies,Paper,Xerox 1906,212.640,6.0,0.0,99.9408,2015-12-26,2010-01-01
9101,9102,CA-2015-163181,11/7/2015,1/1/2010,NaN,AB-10105,Adrian Barton,Consumer,United States,Houston,...,OFF-AR-10001683,Office Supplies,Art,Lumber Crayons,23.640,3.0,0.2,5.3190,2015-11-07,2010-01-01
10010,6753,CA-2016-154767,6/28/2016,1/1/2010,Second Class,BP-11155,NaN,Consumer,United States,Paterson,...,OFF-PA-10003039,Office Supplies,Paper,Xerox 1960,61.960,2.0,0.0,27.8820,2016-06-28,2010-01-01
10033,9240,CA-2017-110310,10/27/2017,1/1/2010,NaN,NB-18655,NaN,Corporate,United States,Tallahassee,...,OFF-PA-10001685,Office Supplies,Paper,Easy-staple paper,56.784,7.0,0.2,20.5842,2017-10-27,2010-01-01


8 lignes détectées (impossible temporellement)

=> Décision : ces lignes seront supprimées dans la phase de nettoyage, conformément aux règles de gestion du cahier des charges.

**Country** : 1 valeur unique ("United States") sur 10 064 lignes.
→ Colonne constante, à supprimer.
(Aucune variance → aucune information

Aucune valeur analytique

Réduit la taille du dataset)
**Row ID** : identifiant séquentiel (1 à 9994), aucune sémantique métier.
→ À conserver comme clé technique (PK de `core.orders`).